libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -561,6 +553,68 @@ inline _CUDA_VSTD::uint64_t * barrier_native_handle(barrier<thread_scope_block>
     return reinterpret_cast<_CUDA_VSTD::uint64_t *>(&b.__barrier);
 }
 
+
+_LIBCUDACXX_NODISCARD_ATTRIBUTE _LIBCUDACXX_DEVICE
+inline barrier<thread_scope_block>::arrival_token arrive_tx(
+    barrier<thread_scope_block> & __b,
+    _CUDA_VSTD::ptrdiff_t __arrive_count_update,
+    _CUDA_VSTD::ptrdiff_t __transaction_count_update) {
+    _LIBCUDACXX_DEBUG_ASSERT(__arrive_count_update >= 0, "Arrival count update must be non-negative.");
+    _LIBCUDACXX_DEBUG_ASSERT(__transaction_count_update >= 0, "Transaction count update must be non-negative.");
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#contents-of-the-mbarrier-object
+    _LIBCUDACXX_DEBUG_ASSERT(__transaction_count_update <= (1 << 20) - 1, "Transaction count update cannot exceed 2^20 - 1.");
+    barrier<thread_scope_block>::arrival_token __token = {};
+    NV_DISPATCH_TARGET(
+        NV_PROVIDES_SM_90, (
+            // XXX: Consider simplifying the logic here. By doing the if else
+            // and trapping, we prevent debugging tools like memcheck/racecheck
+            // to detect that we are passing a pointer with the wrong state
+            // space to mbarrier.arrive. Perhaps it is better to just pass the
+            // pointer directly and let the error handling be done at the
+            // PTX/hardware level.
+            if (__isShared(barrier_native_handle(__b))) {
+                if (__arrive_count_update == 1) {
+                    asm volatile(
```

Comment:
None of these needs to be `volatile`, the `memory` clobber suffices.

Reply:
Changed. Is there a downside to making them volatile?